### ISSUE TITLE
`Arguments for ParamSpec "P@..." are missing` when calling function without wrapper

### ISSUE THREAD

**Issue by samuelcolvin** [microsoft/pyright#8146]
The following code fails pyright with `15:28 - error: Arguments for ParamSpec "P@Foo" are missing`

```py
from typing import Concatenate, ParamSpec, Callable, Generic, Any

P = ParamSpec('P')


class Foo(Generic[P]):
    def __init__(self, func: Callable[Concatenate[int, P], str]) -> None:
        self.f = func

    def call_direct(self, *args: P.args, **kwargs: P.kwargs) -> str:
        return self.f(1, *args, **kwargs)

    def call_compute_args(self) -> str:
        kwargs = self.get_kwargs()
        return self.f(2, **kwargs)

    def get_kwargs(self) -> dict[str, Any]:
        return {'x': 1}
```

I wonder if this is a bug with pyright? If not, is there anyway to correctly type this kind of pattern?

**Comment by samuelcolvin** [comment]
(Worth noting that in my real code `get_kwargs` is doing clever logic to construct the kwargs for `self.f` from its signature).

I can work around the error with:

```py
from typing import Concatenate, ParamSpec, Callable, Generic, Any

P = ParamSpec('P')


class Foo(Generic[P]):
    def __init__(self, func: Callable[Concatenate[int, P], str]) -> None:
        self.f = func

    def call_direct(self, v: int, *args: P.args, **kwargs: P.kwargs) -> str:
        return self.f(v, *args, **kwargs)

    def call_compute_args(self) -> str:
        kwargs = self.get_args()
        return self.call_direct(2, **kwargs)

    def get_args(self) -> dict[str, Any]:
        return {'x': 1}
```

But it seems a bit odd that that's the only way.

**Comment by erictraut** [comment]
This isn't a bug. Pyright is correct to generate an error here. For details, refer to [this section of the typing spec](https://typing.readthedocs.io/en/latest/spec/generics.html#id5). 

Mypy doesn't emit an error for this code, and that's due to a known bug in mypy that hasn't been fixed.

The type of `self.f` is `(int, *args: P.args, **kwargs: P.kwargs) -> str`, which means you must pass a set of positional and keyword arguments that correspond to `P`.

If your intent is for `f` to accept an arbitrary set of keyword-only arguments but no positional arguments (other than the initial `int`), then a `ParamSpec` may not be the right tool. There's no way to limit a `ParamSpec` signature to keyword-only parameters.

If you want to use a `ParamSpec`, you would probably need to use a `cast` to get the type checker to ignore the type violation.

```python
return cast(Callable[..., str], self.f)(2, **kwargs)
```

**Comment by samuelcolvin** [comment]
Thanks for the response.

I need `ParamSpec` for `call_direct` to work properly.

I can generator `*args` as well as `**kwargs` in  in `get_args` or equivalent, that's not the problem.

My problem I guess is that there's no way to do the equivalent of:

```py
args: P.args = self.get_args(...)
kwargs: P.kwargs = self.get_kwargs(...)
self.f(v, *args, **kwargs)
```

Except I guess the route I used in my second comment of calling `self.call_direct` instead of `self.f` directly.

**Comment by erictraut** [comment]
Yeah, if you can generate both `*args` and `**kwargs`, then your approach would be type safe, but I don't see a way to convince a type checker of that fact without using a `cast` — at least within the limitations of today's type system. If this use case is important to you, you could propose an extension or modification to the type system.

**Comment by samuelcolvin** [comment]
It's important, but internal so I'm happy with `# type: ignore` in this scenario.
